src/CoreWCF.Primitives/src/CoreWCF/IdentityModel/Selectors/UserNameSecurityTokenAuthenticator.cs

```diff
@@ -15,13 +16,13 @@ protected UserNameSecurityTokenAuthenticator()
 
         protected override bool CanValidateTokenCore(SecurityToken token)
         {
-            return token is USerNameSecurityToken;
+            return token is UserNameSecurityToken;
         }
 
-        protected override ReadOnlyCollection<IAuthorizationPolicy> ValidateTokenCore(SecurityToken token)
+        protected override ValueTask<ReadOnlyCollection<IAuthorizationPolicy>> ValidateTokenCoreAsync(SecurityToken token)
         {
-            USerNameSecurityToken userNameToken = (USerNameSecurityToken)token;
-            return ValidateUserNamePasswordCore(userNameToken.UserName, userNameToken.Password);
+            UserNameSecurityToken userNameToken = (UserNameSecurityToken)token;
+            return new ValueTask<ReadOnlyCollection<IAuthorizationPolicy>>(ValidateUserNamePasswordCore(userNameToken.UserName, userNameToken.Password));
         }
```

Comment:
Create new virtual method ValidateUserNamePasswordCoreAsync which calls ValidateUserNamePasswordCore and returns the result wrapped in a ValueTask. Make ValidateUserNamePasswordCore virtual and throw NotImplementedException. Put Obsolete attribute on ValidateUserNamePasswordCore saying to implement ValidateUserNamePasswordCoreAsync. This method can just return the ValueTask returned from ValidateUserNamePasswordCoreAsync as it's the only async call and it's the last one in the method. This class is often implemented to look up the username/password in a SQL database and has been a pain point for blocking a thread when doing something which is async so I know this will help some companies.

Reply:
done